src/zon.zig

```diff
@@ -0,0 +1,831 @@
+                    const int_info = res_ty.intInfo(self.sema.pt.zcu);
+                    if (!val.fitsInTwosComp(int_info.signedness, int_info.bits)) {
+                        return self.fail(
+                            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+                            "type '{}' cannot represent integer value '{}'",
+                            .{ res_ty.fmt(self.sema.pt), val },
+                        );
+                    }
+                }
+
+                return self.sema.pt.zcu.intern_pool.get(gpa, self.sema.pt.tid, .{ .int = .{
+                    .ty = res_ty.toIntern(),
+                    .storage = .{ .big_int = val },
+                } });
+            },
+        },
+        .float_literal => |val| {
+            // Check for fractional components
+            if (@rem(val, 1) != 0) {
+                return self.fail(
+                    .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+                    "fractional component prevents float value '{}' from coercion to type '{}'",
+                    .{ val, res_ty.fmt(self.sema.pt) },
+                );
+            }
+
+            // Create a rational representation of the float
+            var rational = try std.math.big.Rational.init(self.sema.arena);
+            rational.setFloat(f128, val) catch |err| switch (err) {
+                error.NonFiniteFloat => unreachable,
+                error.OutOfMemory => return error.OutOfMemory,
+            };
+
+            // The float is reduced in rational.setFloat, so we assert that denominator is equal to
+            // one
+            const big_one = std.math.big.int.Const{ .limbs = &.{1}, .positive = true };
+            assert(rational.q.toConst().eqlAbs(big_one));
+
+            // Check that the result is in range of the result type
+            const int_info = res_ty.intInfo(self.sema.pt.zcu);
+            if (!rational.p.fitsInTwosComp(int_info.signedness, int_info.bits)) {
+                return self.fail(
+                    .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+                    "float value '{}' cannot be stored in integer type '{}'",
+                    .{ val, res_ty.fmt(self.sema.pt) },
+                );
+            }
+
+            return self.sema.pt.zcu.intern_pool.get(gpa, self.sema.pt.tid, .{
+                .int = .{
+                    .ty = res_ty.toIntern(),
+                    .storage = .{ .big_int = rational.p.toConst() },
+                },
+            });
+        },
+        .char_literal => |val| {
+            const rhs: u32 = val;
+            // If our result is a fixed size integer, check that our value is not out of bounds
+            if (Type.zigTypeTag(res_ty, self.sema.pt.zcu) == .int) {
+                const lhs_info = res_ty.intInfo(self.sema.pt.zcu);
+                // If lhs has less than 64 bits, we bounds check. We check at 64 instead of 32 in
+                // case LHS is signed.
```

Comment:
Oh, I'm pretty sure this can actually be at 32, because Unicode codepoint literals always fit in a `u21` -- this union field having type `u32` is a mistake I made in ZOIR, my bad.

Reply:
No worries! I updated Zoir to store a u21 here, and also update this code to reflect that.